pkg/scheduler/framework/interface.go

```diff
@@ -608,6 +613,31 @@ type Handle interface {
 	Parallelizer() parallelize.Parallelizer
 }
 
+// PreFilterResult wraps needed info for scheduler framework to act upon PreFilter phase.
+type PreFilterResult struct {
+	// The set of nodes that should be considered downstream; if null or empty then
```

Comment:
> or empty

Non-nil result with empty NodeNames (no matter it's aggregated, or simply returned by a single PreFilterPlugin) actually denotes "no nodes to continue in downstream", no?

Reply:
No, it means all as well. No nodes is only designated by an unschedulable status.

Reply:
what if pluginA returns {node1, node2}, while pluginB returns {node3, node4}. Let's suppose their logic are both correct, in this case the merging result is an empty obj, and then we're going to iterate all nodes?

Reply:
In the current implementation, yes, that is what we will end up with unless we want RunPreFilterPlugins to detect the case and return an unschedulable status, which I think isn't really worth it.

I am concerned about distinguishing between nil and empty, we already have status that gives us a clear indication whether or not to continue downstream.

Reply:
May be it is worth to do it in RunPreFilterPlugins, I will send a patch and we can check how it looks.

Reply:
what unschedulable message do we report? I still think it is fine to just fallback to the filter phase in this case , we are already doing a full iteration over all nodes just to produce a message when a prefilter returns unschedulable: https://github.com/kubernetes/kubernetes/blob/b4f7da1ec8e01700924a1ab60cf7d12b5af8e3bb/pkg/scheduler/scheduler.go#L914

Reply:
Yeah, I got that; that is what i meant by checking in RunPreFilterPlugins(); the question is what should the unschedulable message say? But more importantly, wouldn't that prevent the pod from going through preemption?

Reply:
I would like to leave it like this, it doesn't seem worth the added quirks in RunPreFilterPlugins.

Reply:
The second commit now implements this. Changed the NodeNames semantics, nil means all nodes while empty means no nodes eligible.

Reply:
A third commit implements the NodeAffinity filter. I guess we can do in the same PR but different commits (I will squash the first two before merging)